File: part 5 - operators precedence/practice/Lexer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace test.part_5.practice
{
    public class Lexer
    {
        public int Position { get; set; }
        string Text { get; init; }
        int Len { get { return this.Text.Length; } }

        public List<Token> Tokens { get; private set; }

        public Lexer(string text)
        {
            Text = text;
            Tokens = new List<Token>();
        }

        private void Error()
        {
            throw new Exception($"Invalid element exception. {Text[Position]} was not expected.");
        }

        private void Skip()
        {
            while (HasNext() && Char.IsWhiteSpace(Text[Position])) { }
        }

        private bool HasNext()
        {
            return ++Position < Len;
        }

        private Token TokenizeInt()
        {
            string number = Text[Position].ToString();

            while (Position < Len - 1 && Char.IsDigit(Text[Position + 1]))
            {
                number += Text[++Position];
            }

            return new Token(Token.INT, number);
        }

        public void Tokenize()
        {
            while (Position < Len)
            {
                switch (Text[Position])
                {
                    case ' ':
                        Skip();
                        continue;

                    case '+':
                        Tokens.Add(new Token(Token.PLUS, "+"));
                        break;

                    case '-':
                        Tokens.Add(new Token(Token.MINUS, "-"));
                        break;

                    case '*':
                        Tokens.Add(new Token(Token.MULTIPLY, "*"));
                        break;

                    case '/':
                        Tokens.Add(new Token(Token.DIVIDE, "/"));
                        break;

                    case '(':
                        Tokens.Add(new Token(Token.LPAREN, "("));
                        break;

                    case ')':
                        Tokens.Add(new Token(Token.RPAREN, ")"));
                        break;

                    default:
                        if (Char.IsDigit(Text[Position]))
                        {
                            Tokens.Add(TokenizeInt());
                        }
                        else
                        {
                            Error();
                        }
                        break;

                }
                HasNext();
            }
        }
    }
}

File: part 14 - nested scopes/lesson/Lexer.py
from Token import RESERVED_KEYWORDS, Token


class Lexer(object):
    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.current_char = self.text[self.pos]

    def error(self):
        raise Exception('Invalid character')

    def advance(self):
        """Advance the `pos` pointer and set the `current_char` variable."""
        self.pos += 1
        if self.pos > len(self.text) - 1:
            self.current_char = None
        else:
            self.current_char = self.text[self.pos]

    def peek(self):
        peek_pos = self.pos + 1
        if peek_pos > len(self.text) - 1:
            return None
        else:
            return self.text[peek_pos]

    def skip_whitespace(self):
        while self.current_char is not None and self.current_char.isspace():
            self.advance()

    def skip_comment(self):
        while self.current_char != '}':
            self.advance()
        self.advance()

    def number(self):
        """Return a (multidigit) integer or float consumed from the input."""
        result = ''
        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self.advance()

        if self.current_char == '.':
            result += self.current_char
            self.advance()

            while self.current_char is not None and self.current_char.isdigit():
                result += self.current_char
                self.advance()

            token = Token('REAL_CONST', float(result))
        else:
            token = Token('INTEGER_CONST', int(result))

        return token

    def _id(self):
        """Handle identifiers and reserved keywords"""
        result = ''
        while self.current_char is not None and self.current_char.isalnum():
            result += self.current_char
            self.advance()

        token = RESERVED_KEYWORDS.get(result.upper(), Token(Token.ID, result.upper()))
        return token

    def get_next_token(self):
        """Lexical analyzer (also known as scanner or tokenizer)
        This method is responsible for breaking a sentence
        apart into tokens. One token at a time.
        """
        while self.current_char is not None:
            if self.current_char.isspace():
                self.skip_whitespace()
                continue

            if self.current_char == '{':
                self.advance()
                self.skip_comment()
                continue

            if self.current_char.isalpha():
                return self._id()

            if self.current_char.isdigit():
                return self.number()

            if self.current_char == ':' and self.peek() == '=':
                self.advance()
                self.advance()
                return Token(Token.ASSIGN, ':=')

            if self.current_char == ';':
                self.advance()
                return Token(Token.SEMI, ';')

            if self.current_char == ':':
                self.advance()
                return Token(Token.COLON, ':')

            if self.current_char == ',':
                self.advance()
                return Token(Token.COMMA, ',')

            if self.current_char == '+':
                self.advance()
                return Token(Token.PLUS, '+')

            if self.current_char == '-':
                self.advance()
                return Token(Token.MINUS, '-')

            if self.current_char == '*':
                self.advance()
                return Token(Token.MUL, '*')

            if self.current_char == '/':
                self.advance()
                return Token(Token.FLOAT_DIV, '/')

            if self.current_char == '(':
                self.advance()
                return Token(Token.LPAREN, '(')

            if self.current_char == ')':
                self.advance()
                return Token(Token.RPAREN, ')')

            if self.current_char == '.':
                self.advance()
                return Token(Token.DOT, '.')

            self.error()

        return Token(Token.EOF, None)
